### PR TITLE
test

### PR DESCRIPTION
**Pull request by sp3c1**
refs #6

### CODE REVIEW

**Review by sp3c1 — COMMENTED** [review]
## Summary
This PR moves the reviewer toward the Pi/CodeGraph architecture: adds a unified-diff context/valid-line map, structured Pi JSON review contract, inline-comment validation, Pi provider flags, and updated docs/env guidance.

## Affected areas / CodeGraph
CodeGraph is available and indexed the Go entry points around `main`, `BuildDiffContext`/`ValidateReviewComments`, and `generateReviewWithPi`. `codegraph affected` did not report affected tests for `main.go`, `diff_context.go`, or `pi_reviewer.go`.

## Risks
- Inline comments may still fail GitHub API validation unless line-based comments include `side: RIGHT` or use the parsed diff position.
- The provided Pi/CodeGraph adapter starts fresh Docker containers instead of using the already-running sandbox, which can lose Pi auth/CodeGraph state and violates the documented runtime assumption.
- CI status is unknown. I attempted `go test ./...`, but `go` is not installed in this environment.

## Tests to run
- `go test ./...`
- `go run . -owner=<owner> -repo=<repo> -pr=<number> -dry -provider=pi`
- Repeat dry run with `-codegraph=false` to verify graceful degradation.

**Note:** This is a self-requested change.

File: pi_reviewer.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+Rules:
+- Inline comments may target ONLY the valid added lines listed in valid_line_table.
+- Do not invent line numbers.
+- Use CodeGraph affected scope to explain blast radius, affected tests, callers/callees, and risk.
+- Prefer recommendation "comment" for author/self-review style output unless there is a concrete blocker.
+- If CI/test status is unknown, mention tests to run instead of claiming tests passed.
+
+Input JSON:
+` + string(data), nil
```
Comment (sp3c1) [inline pi_reviewer.go:161]: [bug] This starts a fresh `docker run` container for review generation, but the project runtime assumptions require using the already-running `pi-sandbox` container via `docker exec`. Starting a new container per review can miss the authenticated Pi/CodeGraph state and conflicts with the documented workflow; switch this adapter and the CodeGraph helper to exec into the stable container instead.

**Review by sp3c1 — COMMENTED** [review]
## Summary
Adds diff grounding, valid inline-comment validation, Pi-based review generation, CodeGraph-aware prompt context, and documentation for the Dockerized Pi workflow.

## Affected areas
- GitHub review generation flow in `main.go`.
- Unified diff parsing and comment validation in `diff_context.go`.
- Pi/CodeGraph provider integration in `pi_reviewer.go`.
- Reviewer setup/runtime docs in `README.md`, `.env.example`, `AGENTS.md`, and `PLAN.md`.

## Risks
- The Pi Docker adapter is the highest-risk path because it controls whether review generation receives the intended prompt/context and whether runtime assumptions match the project’s sandbox model.
- Inline comment validation is a major safety improvement, but edge cases like renames, deleted files, binary files, and GitHub `line` vs `position` compatibility should stay covered by tests.

## CodeGraph affected scope
CodeGraph was available and synced. It reported no affected test files for `pi_reviewer.go` or `diff_context.go`. Related symbols include `generateReviewWithPi`, `ValidateReviewComments`, `IsValidInlineTarget`, `parseHunkHeader`, and `main`, so the blast radius is mainly review generation, review posting safety, and diff-line validation.

## Tests to run
- `go test ./...`
- A dry run against a real PR: `go run . -owner=<owner> -repo=<repo> -pr=<number> -dry`
- A Pi-container integration dry run with CodeGraph enabled and disabled.

Note: I could not run `go test ./...` in this environment because the `go` binary was unavailable.

**Note:** This is a self-requested change.

File: pi_reviewer.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+		Author:         pr.GetUser().GetLogin(),
+		BaseSHA:        pr.GetBase().GetSHA(),
+		HeadSHA:        pr.GetHead().GetSHA(),
+		ValidLineTable: diffCtx.ValidLineSummary(),
+		CodeGraph:      codeGraphContext,
+	}
+	for _, file := range files {
+		if file == nil || file.Filename == nil {
+			continue
+		}
```
Comment (sp3c1) [inline pi_reviewer.go:134]: [risk] The prompt payload should include related issue descriptions for PR bodies like `refs #6`; otherwise the Pi review cannot satisfy the issue-context acceptance goal.

File: pi_reviewer.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+	cmd.Stderr = &stderr
+	if err := cmd.Run(); err != nil {
+		return "", fmt.Errorf("dockerized pi failed: %w\nstderr:\n%s", err, stderr.String())
+	}
+	return stdout.String(), nil
+}
+
+func buildCodeGraphContext(absRepo string, cfg PiConfig, pr *github.PullRequest, files []*github.CommitFile) string {
+	var b strings.Builder
+	status, err := runToolInDocker(absRepo, cfg, "codegraph", "status")
```
Comment (sp3c1) [inline pi_reviewer.go]: [bug] The full PR prompt is sent on stdin, but `pi -p` receives this short argument as the actual prompt. If Pi does not automatically read stdin here, the model will never see the PR metadata, patches, or valid line table. Pass the prompt as the CLI prompt argument or use a mode that explicitly reads stdin.

File: pi_reviewer.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+	}
+	return b.String()
+}
+
+func runToolInDocker(absRepo string, cfg PiConfig, tool string, args ...string) (string, error) {
+	dockerArgs := []string{
+		"run", "--rm",
+		"-v", fmt.Sprintf("%s:%s:rw", absRepo, cfg.ContainerRepoPath),
+		"-e", "PI_PROJECT_DIR=" + cfg.ContainerRepoPath,
+		"-w", cfg.ContainerRepoPath,
```
Comment (sp3c1) [inline pi_reviewer.go:223]: [risk] Running `codegraph affected` once per file can spawn many bounded Docker/tool calls on large PRs. Add a cap or prefer one `codegraph context ... --max-nodes` call plus selected affected-file queries.

**Review by sp3c1 — COMMENTED** [review]
## Summary
This PR moves `gh-pr-reviewer` toward the Pi + CodeGraph architecture: adds unified diff parsing/valid inline line validation, introduces a strict JSON review contract, documents the Dockerized Pi workflow, and adds tests for diff/comment validation and JSON parsing.

## Related issue / intent
Refs #6: “seeing if we can catch context.” The PR is aligned with that intent, but the review prompt path should ensure related issue descriptions are actually passed through to Pi so issue context can influence the author note.

## Affected areas
- Diff grounding: `diff_context.go`, `diff_context_test.go`
- Pi review adapter and JSON parsing: `pi_reviewer.go`, `pi_reviewer_test.go`
- CLI/provider flow and posting behavior: `main.go`
- Runtime docs/config: `README.md`, `.env.example`, `AGENTS.md`, `PLAN.md`

## Risks / notes
- Main functional risk: the Pi invocation may not receive the full structured PR prompt if `pi -p` does not consume stdin while also receiving a prompt argument.
- CodeGraph scope from my local check: CodeGraph is initialized but reports pending changes; `codegraph affected` reported no affected test files for `diff_context.go`, `pi_reviewer.go`, or `main.go`.
- Local tests could not be run here because `go` is not installed in this environment.

## Tests to run
- `go test ./...`
- Dry-run against this PR with Pi container running: `go run . -owner=nvrwhr -repo=gh-pr-reviewer -pr=<number> -dry -forcedry`
- Verify generated review JSON contains valid inline comments only on parsed added lines.
- Verify CodeGraph unavailable/uninitialized path degrades with a warning, not failure.

**Note:** This is a self-requested change.
